Adds NTT Pipeline, optimized Dahlia CT.

cgyurgyik opened the pull request:
In this PR there is:
1. Re-hauled NTT directory, placing Dahlia implementations in its own corner for now.
2. Prototype pipeline generator for FuTIL (written in Python for now).
3. NTT correctness for FuTIL pipeline with N=32, generation for N=4

rachitnigam commented on ntt/naive/ntt.expect:
```diff
@@ -0,0 +1,3 @@
+{
+  "cycles": 65241
```
Wow, that's a lot of cycles. On your machine, how much time (in seconds) did it take to run this?

cgyurgyik replied:
~1 minute

rachitnigam commented:
the `--pass-debug` flag to the dahlia compiler prints the program after every pass in the Dahlia compiler. Might be useful. Especially the pass called "sequentialize".

cgyurgyik commented:
I've re-hauled this a bit. To prevent a catastrophic drop in developer productivity because NTT tests are taking too long in the automated test suite, I've added a single correctness test for the FuTIL pipeline with N=32. With #361, we can later add cycle count as well.

The Dahlia implementations are placed in a separate directory for now. 

Also added is the prototype Python pipeline generator.

rachitnigam commented:
I don't think we should commit the large FuTIL files. As with other frontends, create two kinds of tests:
1. Frontend generation tests: These should be small and only test the output FuTIL files
2. Correctness tests: These can be larger tests but they expect files should test the outputs and not the generated FuTIL code.

cgyurgyik commented:
> I don't think we should commit the large FuTIL files. As with other frontends, create two kinds of tests:
> 
> 1. Frontend generation tests: These should be small and only test the output FuTIL files
> 2. Correctness tests: These can be larger tests but they expect files should test the outputs and not the generated FuTIL code.

- Removed ntt-256, ntt-512. 
- I made the Python generator read in a file that takes in the 3 values for generation (without much error checking). There is now a generation test as well.

rachitnigam commented:
Can you move the front end tests to the tests directory (same as systolic array)? I think in the long term we’ll probably add NTT as a fud stage and make it a first-class front end.

cgyurgyik commented:
> Can you move the front end tests to the tests directory (same as systolic array)? I think in the long term we’ll probably add NTT as a fud stage and make it a first-class front end.

Done. I also moved the NTT folder to frontends for now. I just saw `gen-systolic-array.py` so I shamelessly copied the argument parsing. If this script stays awhile longer, I'll look into refactoring the two since they share a lot of codegen features. We should discuss next steps for this on Tuesday as well.

cgyurgyik commented:
Hmm looks like some unrelated clippy failures
```
error[E0119]: conflicting implementations of trait `std::convert::From<ir::guard::Guard>` for type `boolean_expression::Expr<ir::guard::Guard>`:
Error:   --> calyx/src/passes/simplify_guards.rs:10:1
   |
10 | impl From<ir::Guard> for Expr<ir::Guard> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `boolean_expression`:
           - impl<T> From<T> for boolean_expression::Expr<T>
             where T: Clone, T: Debug, T: Eq, T: Hash;
```

rachitnigam commented:
The problem is https://github.com/cfallin/boolean_expression/issues/16.  Fixed on master by pinning working version of the library.

rachitnigam commented on tests/correctness/ntt-pipeline/ntt-32.futil:
```diff
@@ -0,0 +1,4167 @@
+// +----+-------------------------+-------------------------+-------------------------+--------------------------+--------------------------+
```
Sorry for yet another back-and-forth about this. Can we commit smaller examples as tests? The use case of these test files is that they are *human-readable*. When a change occur, the person committing the change should actually be able to read the file and see if something unexpected happened. No one will ever read a 4,000 line file.

rachitnigam replied:
Ah, I just saw `ntt-4` down below. My expectation with the correctness test is that you don't actually commit the FuTIL file. The frontend first generates the file and then runs it using `fud`. Committing the FuTIL files has the downside that correctness tests don't remain in sync with changes to the frontend.

cgyurgyik replied:
No problem. That makes sense. I just went with ntt-32 for the correctness test since that's what the original naive method had. I'm going to be on the road today but I'll fix this later. I presume a few smaller correctness tests is viable as well.

cgyurgyik replied:
Update: I've removed then ntt-32 and added ntt-16, ntt-8 tests for correctness, with their cycle counts. These are now generated directly from the python script using a temporary file.

 I've left the ntt-4 for the frontend generation test, since I think ntt-2 is a bit too trivial.

Let me know if I'm missing anything.

rachitnigam commented:
LGTM!
